### ISSUE TITLE
Unable to self-assign issues with jira cli

### ISSUE THREAD

**Issue by pcockwell** [go-jira/jira#314]
I've been using the jira cli as part of custom commands I run on my system to simultaneously create a branch based on the ticket name, assign it to myself in Jira, and move the issue to In Progress. Until recently, it has been working quite well. I was on vacation for a few weeks, and upon my return it has stopped working.

- Whenever I run `jira view TICKET_ID`, it returns the ticket details fine
- Whenever I run `jira take TICKET_ID`, I get the following error:
```
ERROR Invalid Usage: 'accountId' must be the only user identifying query parameter in GDPR strict mode.
```

I have been playing around with my API tokens (which are stored in `keyring` on a Macbook), and `.jira.d/config.yml` to see if I can isolate the cause and, along with the error message, have determined the following:
- The new system JIRA is rolling out for GDPR requires using the `accountId` field as a replacement of the `user` field for requests. I have since updated my config to include my `accountId`
- The jira cli will not allow a login using an API token without at least one of the `user` or `login` fields
  - If I remove this field from my config, it incorrectly registers which user I am (thus implying that the `accountId` field is not working for identification purposes)

My current `.jira.d/config.yml` contains the following fields (that have been anonymized):
```
endpoint: https://mycompany.atlassian.net                                                                                                                                      
accountId: MY_PROFILE_ACCOUNT_ID
user: patrick@mycompany.com
password-source: keyring
```

My guess is that the `user` and `login` fields are also the culprit to the GDPR based message, simply because it is complaining that `user identifying query parameter`s are being passed in to the API request.

Is someone able to explain what I am doing wrong, and/or the correct fix for this? I really really like my personal tooling as it increases my productivity, and decreases the repetitive mundane tasks I need to do on a regular basis, and would really appreciate being able to use them again.

Thanks

**Comment by dvdbng** [comment]
Same here, it was working last Friday so it must have been changes in jira API

**Comment by pcockwell** [comment]
@dvdbng I've made an attempt at a fix for this, in the PR above. I've tested the functionality (assign/watch) locally, and it works. Now, we wait on maintainers to see how poor my code is, and what testing related changes need to be made haha. If you are urgently in need of the functionality, you can use those changes, and run `make build && make install` locally. This will give you a `jira` binary located at `$HOME/bin/jira`.

**Comment by keien** [comment]
I just ran into this as well and did the above - seems to work, though I struggled with figuring out what I was supposed to have in my `config.yml` now. The PR looked like I'm supposed to add a new `account-id` field, but I tried various combinations of `user`, `login`, and `account-id`, and in the end just keeping my initial `user` and `login` and no `account-id` seems to work, whereas if I add `account-id` it gives me permission errors. Maybe I'm misunderstanding the fix and how it applies to the requests?

**Comment by keien** [comment]
Just kidding - with no `account-id`, the `take` command returns an OK but doesn't do anything, and `assign` still says `ERROR Invalid Usage: Specified user does not exist or you do not have required permissions`

**Comment by pcockwell** [comment]
@keien The way mine is setup is that I have a `login` field so that the login functionality still works, and I have an `account-id` field so that the `assign`, `take`, and `watch` functionality works (with the changes).

**Comment by dhollinger** [comment]
Related: https://developer.atlassian.com/cloud/jira/platform/deprecation-notice-user-privacy-api-migration-guide/

They removed `username` and `userkey` from the API entirely. You HAVE to now use `accountId` for everything

**Comment by keien** [comment]
hmm, maybe I'm doing something wrong but I'm not getting it to work even with the fix
```
$ cd $GOPATH/src/github.com/go-jira/jira
$ git log -1
commit 148330885dc411076b737a200a835e8177be83d6
Author: Patrick Cockwell <pcockwell@gmail.com>
Date:   Tue Feb 18 11:32:07 2020 +0700

    Add account-id parameter to config parsing; Use account-id parameter instead of user or name for watch, and assign issue commands      
$ cat .jira.d/config.yml
endpoint: https://XXXXXX.atlassian.net
login: keien@XXXXXXXX
account-id: keien@XXXXXXX
$ sha256sum ~/bin/jira
f642d637d5e8b7de98cf636ccdad760c909a07c1d65aadcfcad76ba4722d614d  /home/keien/bin/jira
```

**Comment by dhollinger** [comment]
@keien I ended up pulling down @pcockwell's fix branch (NOT his master branch), built the binary and placed it in my `~/bin` directory and it works just fine..... as long as I'm NOT in the go-jira source directory

**Comment by keien** [comment]
I also have maybe a related issue where all of the user names in `jira view` show up as `<no value>` now

**Comment by dhollinger** [comment]
You have to change any template using `fields.assignee.name` to `fields.assignee.displayName`

**Comment by pcockwell** [comment]
@keien Your `accountId` is not the email address. It's a unique key assigned to your account by JIRA/Atlassian. You'll need to select your profile icon on the bottom left, select `Profile`, and then select the characters at the end of the URL. For example, my profile URL is:

```
https://flyrlabs.atlassian.net/jira/people/5acdc3181f4cab2b635489d7
```
This means that my `accountId` is `5acdc3181f4cab2b635489d7`.

**Comment by pcockwell** [comment]
@dhollinger @keien I just pushed another commit per the comments @dhollinger made regarding using `displayName` instead of `name`. This is now also included on my branch.

**Comment by keien** [comment]
@pcockwell thanks, didn't know that; filling in my real account ID worked

for `displayName`, seems with the latest from your branch, fields like assignee and reporter are now showing up, but names for comments are not, yet

**Comment by pcockwell** [comment]
> @pcockwell thanks, didn't know that; filling in my real account ID worked
> 
> for `displayName`, seems with the latest from your branch, fields like assignee and reporter are now showing up, but names for comments are not, yet

I'll take a look at that

**Comment by pcockwell** [comment]
Fixed. Unfortunately, in cases where someone has been tagged, that is included in the comment body as a reference to the `accountId` of that user. As such, there isn't really a way for me to convert that to the actual user's `displayName`

**Comment by keien** [comment]
totally fine, thanks for the quick fixes!

do you know when this stuff will make it into the main repo?

**Comment by pcockwell** [comment]
Hey @keien. Per @coryb in #315:

> Hey, thanks for the PR!  Not sure how to proceed with this, unfortunately, we cannot merge as-is.  This is super frustrating that Atlassian has broken their APIs while not bothering to change the API versions.  One major issue is that this change is only for jira-cloud, but this tool is used a bunch with corporate on-premises Jira installs, so we will need to figure out some sort of compatibility.
> 
> We can usually auto-detect jira-cloud since it uses atlassian.net hostnames, but ideally, we would be able to interrogate a Jira API to get information about how the client needs to behave, not sure if something like this exists for Jira.
> 
> Also ideally we would be able to figure out what the `account-id` property should be from the login information without having yet another identification property in the config file.  
> 
> If anyone has a lead on any of these topics I would appreciate the help.  I think this will require some research to figure out how to solve the compatibility issues, and research takes time which I am very short on these days.
> 
> _Originally posted by @coryb in https://github.com/go-jira/jira/pull/315#issuecomment-589997719_

**Comment by coryb** [comment]
There is a new PR #317 if anyone wants to try out the code.  The changes will try to detect jira-cloud vs jira-server (enterprise on-prem solution).  It will also try to convert any `username` inputs to the `account-id` when translating input to the APIs.  It should work without extra user configuration.  There are many edge cases where users are provided as input, I am not 100% sure I caught them all but at least the tests pass now.

**Comment by coryb** [comment]
I believe the latest release has fixed this:
https://github.com/go-jira/jira/releases/tag/v1.0.23

**Comment by keien** [comment]
@coryb thanks for the update! And thanks to @pcockwell as well for providing the stopgap solution!

**Comment by wari** [comment]
The latest release, as well as the one I compiled myself, does not seem to work for me:
```
$ ./jira-linux-amd64 take DOI-8230
.....
ERROR Invalid Usage: The query parameter 'username' is not supported in GDPR strict mode.
```

**Comment by nyarly** [comment]
I have the same issue; I'm compiling from `v1.0.23`

**Comment by prashankMT** [comment]
same issue with **jira assign** in version **1.0.23**

**Comment by prashankMT** [comment]
@ldelossa Thanks for the update

**Comment by prashankMT** [comment]
@ldelossa any timeline on the release ??
